compiler/hash-typecheck/src/traverse/mod.rs

```diff
@@ -286,30 +290,103 @@ impl<'gs, 'ls, 'cd, 'src> visitor::AstVisitor for TcVisitor<'gs, 'ls, 'cd, 'src>
         &mut self,
-        _ctx: &Self::Ctx,
-        _node: hash_ast::ast::AstNodeRef<hash_ast::ast::RefExpr>,
+        ctx: &Self::Ctx,
+        node: hash_ast::ast::AstNodeRef<hash_ast::ast::RefExpr>,
     ) -> Result<Self::RefExprRet, Self::Error> {
-        todo!()
+        let walk::RefExpr { inner_expr, mutability } = walk::walk_ref_expr(self, ctx, node)?;
+
+        // Depending on the `mutability` of the reference, create the relevant type
+        // function application
+        let ref_def = if mutability.is_some() {
+            self.core_defs().reference_mut_ty_fn
+        } else {
+            self.core_defs().reference_ty_fn
+        };
+
+        let builder = self.builder();
+
+        // Create either `RefMut<T>` or `Ref<T>`
+        let ref_args =
+            builder.create_args([builder.create_arg("T", inner_expr)], ParamOrigin::TyFn);
+        let ref_ty = builder.create_app_ty_fn_term(ref_def, ref_args);
+
+        let term = builder.create_rt_term(ref_ty);
+
+        // Add location to the type:
+        let ref_expr_span = self.source_location(node.span());
+        self.location_store_mut().add_location_to_target(term, ref_expr_span);
+
+        Ok(term)
     }
 
     type DerefExprRet = TermId;
 
     fn visit_deref_expr(
         &mut self,
-        _ctx: &Self::Ctx,
-        _node: hash_ast::ast::AstNodeRef<hash_ast::ast::DerefExpr>,
+        ctx: &Self::Ctx,
+        node: hash_ast::ast::AstNodeRef<hash_ast::ast::DerefExpr>,
     ) -> Result<Self::DerefExprRet, Self::Error> {
-        todo!()
+        let walk::DerefExpr(inner) = walk::walk_deref_expr(self, ctx, node)?;
+        let inner_ty = self.typer().ty_of_term(inner)?;
+
+        // Create a `Ref<T>` dummy type for unification...
+        let ap_ref_ty = {
+            let ref_ty = self.core_defs().reference_ty_fn;
+            let builder = self.builder();
+
+            builder.create_app_ty_fn_term(
+                ref_ty,
+                builder.create_args(
+                    [builder.create_arg("T", builder.create_unresolved_term())],
+                    ParamOrigin::TyFn,
+                ),
+            )
+        };
+
+        // Create a `RefMut<T>` dummy type for unification...
+        let ap_ref_mut_ty = {
+            let ref_mut_ty = self.core_defs().reference_mut_ty_fn;
+            let builder = self.builder();
+
+            builder.create_app_ty_fn_term(
+                ref_mut_ty,
+                builder.create_args(
+                    [builder.create_arg("T", builder.create_unresolved_term())],
+                    ParamOrigin::TyFn,
+                ),
+            )
+        };
+
+        // Attempt to unify this with a `Ref<T>` to see if the `inner_ty` can
+        // be dereferenced. If that fails, then try to unify with `RefMut<T>`
+        let result = self
+            .unifier()
+            .unify_terms(inner_ty, ap_ref_ty)
+            .or_else(|_| self.unifier().unify_terms(inner_ty, ap_ref_mut_ty))?;
```

Comment:
This is fine, however in the future we might want to keep the first error rather than the second (or even have a custom error)

Reply:
I'll add a to-do note